src/util/bip152.rs

```diff
@@ -0,0 +1,335 @@
+
+        let mut prefilled = vec![];
+        let mut short_ids = vec![];
+        let mut last_prefill = 0;
+        for (idx, tx) in block.txdata.iter().enumerate() {
+            if idx == 0 || prefill[idx - 1] {
+                let diff_idx = idx - last_prefill;
+                last_prefill = idx;
+                prefilled.push(PrefilledTransaction(
+                    diff_idx as u64,
+                    match version {
+                        1 => {
+                            // strip witness for version 1
+                            let mut no_witness = tx.clone();
+                            no_witness.input.iter_mut().for_each(|i| i.witness.clear());
+                            no_witness
+                        }
+                        2 => tx.clone(),
+                        _ => unreachable!(),
+                    },
+                ));
+            } else {
+                short_ids.push(ShortId::with_siphash_keys(
+                    &match version {
+                        1 => tx.txid(),
+                        2 => tx.bitcoin_hash(),
+                        _ => unreachable!(),
+                    },
+                    siphash_keys,
+                ));
+            }
+        }
+
+        Ok(HeaderAndShortIds {
+            header: block.header.clone(),
+            nonce: nonce,
+            // Provide coinbase prefilled.
+            prefilled_txs: prefilled,
+            short_ids: short_ids,
+        })
+    }
+}
+
+/// A BlockTransactionsRequest structure is used to list transaction indexes
+/// in a block being requested.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct BlockTransactionsRequest {
+    ///  The blockhash of the block which the transactions being requested are in.
+    pub block_hash: sha256d::Hash,
+    ///  The indexes of the transactions being requested in the block.
+    pub indexes: Vec<u64>,
+}
+
+impl Encodable for BlockTransactionsRequest {
+    fn consensus_encode<S: io::Write>(&self, mut s: S) -> Result<usize, encode::Error> {
+        let mut len = self.block_hash.consensus_encode(&mut s)?;
+        // Manually encode indexes because they are differentially encoded VarInts.
+        len += VarInt(self.indexes.len() as u64).consensus_encode(&mut s)?;
+        let mut last_idx = 0;
+        for idx in &self.indexes {
+            len += VarInt(*idx - last_idx).consensus_encode(&mut s)?;
+            last_idx = *idx;
+        }
+        Ok(len)
+    }
+}
+
+impl Decodable for BlockTransactionsRequest {
+    fn consensus_decode<D: io::Read>(mut d: D) -> Result<BlockTransactionsRequest, encode::Error> {
+        Ok(BlockTransactionsRequest {
+            block_hash: sha256d::Hash::consensus_decode(&mut d)?,
+            indexes: {
+                // Manually decode indexes because they are differentially encoded VarInts.
+                let differential: Vec<VarInt> = Decodable::consensus_decode(&mut d)?;
```

Comment:
Not going to block this PR on it, but this could be done much more efficiently by avoiding the temporary `Vec`.

Reply:
Done in separate commit.